sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala

```diff
@@ -85,12 +85,24 @@ case class DecimalType(precision: Int, scale: Int) extends FractionalType {
   private[sql] def isWiderThan(other: DataType): Boolean = other match {
     case dt: DecimalType =>
       (precision - scale) >= (dt.precision - dt.scale) && scale >= dt.scale
-    case dt: IntegralType =>
```

Comment:
`forType` returns the default DecimalType for primitive, that could have smaller range than it, for example, Float has larger range than DecimalType(24, 7).